rhino/src/main/java/org/mozilla/javascript/optimizer/BodyCodegen.java

```diff
@@ -1370,6 +1370,50 @@ private void generateExpression(Node node, Node parent) {
                 }
                 break;
 
+            case Token.GETELEM_OPTIONAL:
+                {
+                    generateExpression(child, node); // object
+
+                    int getExpr = cfw.acquireLabel();
+                    int putUndefined = cfw.acquireLabel();
+                    int after = cfw.acquireLabel();
+
+                    // if (object == null) goto putUndefined
+                    cfw.add(ByteCode.DUP);
+                    ++maxStack;
+                    cfw.add(ByteCode.IFNULL, putUndefined);
+
+                    // if (!Undefined.isUndefined(object)) goto getExpr
+                    cfw.add(ByteCode.DUP); // No need to increase maxStack again!
+                    cfw.addInvoke(
+                            ByteCode.INVOKESTATIC,
+                            "org.mozilla.javascript.Undefined",
+                            "isUndefined",
+                            "(Ljava/lang/Object;)Z");
+                    cfw.add(ByteCode.IFEQ, getExpr);
+
+                    cfw.markLabel(putUndefined);
+                    cfw.add(ByteCode.POP); // Remove object
+                    cfw.add(
+                            ByteCode.GETSTATIC,
+                            "org/mozilla/javascript/Undefined",
+                            "instance",
+                            "Ljava/lang/Object;");
+                    cfw.add(ByteCode.GOTO, after);
+
+                    cfw.markLabel(getExpr);
+                    generateExpression(child.getNext(), node); // id
+                    cfw.addALoad(contextLocal);
+                    cfw.addALoad(variableObjectLocal);
+                    if (node.getIntProp(Node.ISNUMBER_PROP, -1) != -1) {
+                        addDynamicInvoke("PROP:GETINDEX", Signatures.PROP_GET_INDEX);
```

Comment:
Awesome -- using the new instructions!

I do see from the coverage report that we don't get to this specific case, and given that we just changed it, I'd feel better if we did -- I have a suggestion below.

Reply:
I copy&pasted the code from the existing `GETELEM` path [here](https://github.com/mozilla/rhino/blob/master/rhino/src/main/java/org/mozilla/javascript/optimizer/BodyCodegen.java#L1366):

```java
            case Token.GETELEM:
                generateExpression(child, node); // object
                generateExpression(child.getNext(), node); // id
                cfw.addALoad(contextLocal);
                cfw.addALoad(variableObjectLocal);
                if (node.getIntProp(Node.ISNUMBER_PROP, -1) != -1) {
                    addDynamicInvoke("PROP:GETINDEX", Signatures.PROP_GET_INDEX);
                } else {
                    addDynamicInvoke("PROP:GETELEMENT", Signatures.PROP_GET_ELEMENT);
                }
                break;
```

However, after trying to add a simple test case for this, I've actually realized that the _original_ code I copied does not work either. I.e. if I do this:

```java
        Utils.assertWithAllOptimizationLevelsES6(true, "o = {[0]: true}; o[0]");
```

then the code path for `PROP:GETINDEX` is not followed. So, this looks like a problem in https://github.com/mozilla/rhino/pull/1645